src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -0,0 +1,374 @@
+    }
+
+    private List<List<ByteBuffer>> bindAndGetElements(QueryOptions options)
+    {
+        List<List<ByteBuffer>> elementsList = values.bindAndGetElements(options);
+        validate(elementsList);
+        elementsList.forEach(this::validateElements);
+        return elementsList;
+    }
+
+    private void validate(List<?> list)
+    {
+        if (list == null)
+            throw invalidRequest("Invalid null value for %s", columnsExpression);
+        if (list == Term.UNSET_LIST)
+            throw invalidRequest("Invalid unset value for %s", columnsExpression);
+    }
+
+    private <T> void validate(ByteBuffer buffer)
+    {
+        if (buffer == null)
+            throw invalidRequest("Invalid null value for %s", columnsExpression);
+        if (buffer == ByteBufferUtil.UNSET_BYTE_BUFFER)
+            throw invalidRequest("Invalid unset value for %s", columnsExpression);
+    }
+
+    private void validateElements(List<ByteBuffer> elements)
+    {
+        validate(elements);
+
+        List<ColumnMetadata> columns = columns();
+        for (int i = 0, m = columns.size(); i < m; i++)
+        {
+            ColumnMetadata column = columns.get(i);
+            ByteBuffer element = elements.get(i);
+            if (element == null)
+                throw invalidRequest("Invalid null value for %s in %s",
+                                     column.name.toCQLString(), columnsExpression);
+            if (element == ByteBufferUtil.UNSET_BYTE_BUFFER)
+                throw invalidRequest("Invalid unset value for %s in %s",
+                                     column.name.toCQLString(), columnsExpression);
+        }
+    }
+
+    @Override
+    public void addToRowFilter(RowFilter filter, IndexRegistry indexRegistry, QueryOptions options)
+    {
+        if (isOnToken())
+            throw new UnsupportedOperationException();
+
+        switch (columnsExpression.kind())
+        {
+            case SINGLE_COLUMN:
+                List<ByteBuffer> buffers = bindAndGet(options);
+
+                ColumnMetadata column = firstColumn();
+                if (operator == Operator.IN)
+                {
+                    filter.add(column, operator, inValues(column, buffers));
+                }
+                else if (operator == Operator.LIKE)
+                {
+                    LikePattern pattern = LikePattern.parse(buffers.get(0));
+                    // there must be a suitable INDEX for LIKE_XXX expressions
+                    RowFilter.SimpleExpression expression = filter.add(column, pattern.kind().operator(), pattern.value());
+                    indexRegistry.getBestIndexFor(expression)
+                                 .orElseThrow(() -> invalidRequest("%s is only supported on properly indexed columns",
+                                                                   expression));
+                }
+                else
+                {
+                    filter.add(column, operator, buffers.get(0));
```

Comment:
Map element restriction use to be implemented within the Contains restriction which made things even more confusing. :-( 

The IS NOT NULL logic was useless as you could not reach it.

Reply:
True, my issue is with the old code, not the new one. Just trying to find out if we miss some weird corner case, etc.
Otherwise the new one reads perfect!